### PR TITLE
Adding JSON to FBP support

### PR DESCRIPTION
**Pull request by aretecode**
https://github.com/flowbased/fbp/issues/52

### CODE REVIEW

File: package.json [hunk 1]
```diff
@@ -1,11 +1,17 @@
 {
   "name": "fbp",
   "description": "Parser for the .fbp flow definition language",
-  "version": "1.4.0",
+  "version": "1.5.0",
```
Comment (jonnor) [inline package.json]: Please don't change version numbers in PRs, it very quickly creates merge conflicts. Better to add entry in CHANGES.md

File: lib/json.js [hunk 2]
```diff
@@ -0,0 +1,84 @@
+var indexOf = [].indexOf || function(item) { for (var i = 0, l = this.length; i < l; i++) { if (i in this && this[i] === item) return i; } return -1; };
+
+module.exports = {
+  parse: function(graph, options) {
```
Comment (jonnor) [inline lib/json.js]: The function should take a JS object of the graph, not a string with JSON encoded data.

I think it is misleading to call this a parser. It serializes to the .FBP format. It is the inverse function of `fbp.parse`

File: CHANGES.md [hunk 3]
```diff
@@ -1,3 +1,6 @@
+# fbp 1.5.0 - released 02.07.2016
```
Comment (jonnor) [inline CHANGES.md]: Better to leave as released TODO or similar, the date is already wrong :p

File: lib/json.coffee [hunk 4]
```diff
@@ -0,0 +1,74 @@
+module.exports =
+  parse: (graph, options = {}) ->
```
Comment (jonnor) [inline lib/json.coffee]: Please let the parse function be at the first (no) indentation level. Then at the end do the `module.export = ...`

File: spec/json.coffee [hunk 5]
```diff
@@ -0,0 +1,349 @@
+if typeof process isnt 'undefined' and process.execPath and process.execPath.indexOf('node') isnt -1
+  chai = require 'chai' unless chai
+  jsonParser = require '../lib/json'
+  fbpParser = require '../lib/fbp'
+  fbpParser.validateSchema = true # validate schema for every test on node.js. Don't have tv4 in the browser build
+else
+  fbpParser = require 'fbp'
+  jsonParser = require '../lib/json'
```
Comment (jonnor) [inline spec/json.coffee]: Don't do a special require for the JSON parser. It should be exposed on the `fbp.serialize` (or similar).

Tests must test how external consumers are to use the functionality, and it is not right to expect them to dig inside node_modules/fbp/lib/json...

File: spec/json.coffee [hunk 6]
```diff
@@ -0,0 +1,349 @@
...
+            'ReadTemplate'
+            'Render'
+          ]
+        it 'should contain ten edges and IIPs', ->
+          chai.expect(graphData2.connections).to.be.an 'array'
+          chai.expect(graphData2.connections.length).to.equal 10
+        it 'should contain no exports', ->
+          chai.expect(graphData2.exports).to.be.an 'undefined'
+
+  describe 'JSON to FBP parser', ->
```
Comment (jonnor) [inline spec/json.coffee]: Duplicate description?

**Comment by jonnor** [comment]
Gah, more checked in generated code. Guess we have to as long as we support component.json... :(

File: grammar/fbp.peg [hunk 7]
```diff
@@ -17,6 +17,93 @@
 
   nodes = {};
 
+  var serialize, indexOf = [].indexOf || function(item) { for (var i = 0, l = this.length; i < l; i++) { if (i in this && this[i] === item) return i; } return -1; };
+
+  parser.serialize = function(graph) {
```
Comment (jonnor) [inline grammar/fbp.peg]: Uhm, this code is generated from the CoffeeScript. How does it end up in here during build process?
It is not acceptable to require manual steps to keep this in sync... In general, mixing generated code and not-generated is pretty bad.

File: spec/json.coffee [hunk 8]
```diff
@@ -0,0 +1,341 @@
...
+            'Render'
+          ]
+        it 'should contain ten edges and IIPs', ->
+          chai.expect(graphData2.connections).to.be.an 'array'
+          chai.expect(graphData2.connections.length).to.equal 10
+        it 'should contain no exports', ->
+          chai.expect(graphData2.exports).to.be.an 'undefined'
+
+  describe 'with flowhub json graph (from ingress c-base table)', ->
+    jsonData = """
```
Comment (jonnor) [inline spec/json.coffee]: Fixture not moved ro separate file

File: lib/json.coffee [hunk 9]
```diff
@@ -0,0 +1,75 @@
+serialize = (graph, options = {}) ->
+  input = JSON.parse graph
```
Comment (jonnor) [inline lib/json.coffee]: Still expects string input

File: grammar/fbp.peg (hunk 7 above)
Comment (jonnor) [inline grammar/fbp.peg]: How I think this should be solved is to let `lib/fbp.js ` be a concatenation of
* the JS generated by PEG
* the JS generated from CS
* one liner which sets `module.exports.serialize`

The intermediate .ja files should not be checked into git, only the last result.

Comment (jonnor) [inline grammar/fbp.peg]: Alternative is to not use CoffeeScript at all and just maintain the JS inline in the grammar

File: spec/json.coffee (hunk 8 above)
Comment (aretecode) [inline spec/json.coffee]: I think it is fine having the fixture in here, is it really worth adding a file loader for both js and node just to remove data in a test that is half as long as the other test?

Comment (jonnor) [inline spec/json.coffee]: You've got a point about the browser tests. Silly that it is so tricky to load a file asset there...

**Pull request merged**